eclair-core/src/test/scala/fr/acinq/eclair/payment/relay/NodeRelayerSpec.scala

```diff
@@ -323,6 +327,131 @@ class NodeRelayerSpec extends ScalaTestWithActorTestKit(ConfigFactory.load("appl
+    relayMonitor.expectMessageType[NodeRelay.Command]
+
+    // fail when not triggered after the last incoming part received
+    incomingAsyncPayment.foreach { p =>
+      val fwd = register.expectMessageType[Register.Forward[CMD_FAIL_HTLC]]
+      assert(fwd.channelId == p.add.channelId)
+      assert(fwd.message == CMD_FAIL_HTLC(p.add.id, Right(PaymentTimeout), commit = true))
+    }
+
+    register.expectNoMessage(100 millis)
+  }
+
+  test("relay waiting async payment when triggered") { f =>
+    val g = f.copy(nodeParams = f.nodeParams.copy(features = f.nodeParams.features.add(AsyncPaymentPrototype, Optional), relayParams = f.nodeParams.relayParams.copy(timeout = 5 seconds)))
+    import g._
+
+    val (nodeRelayer, parent) = g.createNodeRelay(incomingAsyncPayment.head)
+
+    incomingAsyncPayment.foreach { p=>
+      nodeRelayer ! NodeRelay.Relay(p)
+      relayMonitor.expectMessageType[NodeRelay.Relay]
+    }
+    mockPayFSM.expectNoMessage(100 millis) // we should NOT trigger a downstream payment before we received a complete upstream payment
+
+    // send trigger after receiving the (private) WrappedMultiPartPaymentSucceeded message
+    relayMonitor.expectMessageType[NodeRelay.Command]
+
+    // increase block height to one block before the outgoing expiry
+    blockHeight.set(outgoingExpiry.toLong - 1)
+
+    // trigger the node relay to forward the payment
+    nodeRelayer ! NodeRelay.RelayAsyncPayment
+
+    // upstream payment relayed
+    val outgoingCfg = mockPayFSM.expectMessageType[SendPaymentConfig]
+    validateOutgoingCfg(outgoingCfg, Upstream.Trampoline(incomingAsyncPayment.map(_.add)))
+    val outgoingPayment = mockPayFSM.expectMessageType[SendMultiPartPayment]
+    validateOutgoingPayment(outgoingPayment)
+    // those are adapters for pay-fsm messages
+    val nodeRelayerAdapters = outgoingPayment.replyTo
+
+    // A first downstream HTLC is fulfilled: we should immediately forward the fulfill upstream.
+    nodeRelayerAdapters ! PreimageReceived(paymentHash, paymentPreimage)
+    incomingAsyncPayment.foreach { p =>
+      val fwd = register.expectMessageType[Register.Forward[CMD_FULFILL_HTLC]]
+      assert(fwd.channelId == p.add.channelId)
+      assert(fwd.message == CMD_FULFILL_HTLC(p.add.id, paymentPreimage, commit = true))
+    }
+
+    // Once all the downstream payments have settled, we should emit the relayed event.
+    nodeRelayerAdapters ! createSuccessEvent()
+    val relayEvent = eventListener.expectMessageType[TrampolinePaymentRelayed]
+    validateRelayEvent(relayEvent)
+    assert(relayEvent.incoming.toSet == incomingAsyncPayment.map(i => PaymentRelayed.Part(i.add.amountMsat, i.add.channelId)).toSet)
+    assert(relayEvent.outgoing.nonEmpty)
+    parent.expectMessageType[NodeRelayer.RelayComplete]
+    register.expectNoMessage(100 millis)
+  }
+
+  test("fail to relay async payment when triggered if the outgoing expiry is not above chain height") { f =>
```

Comment:
We should add one more test that verifies that when the async payment feature is turned off, we instantly relay payments (this is what guarantees that it's safe to merge on `master` since it's off by default).

Reply:
Fixed in [36b0d42](https://github.com/ACINQ/eclair/pull/2435/commits/36b0d421cf820431a4ac9a8bbe54209cf9f5bdea)